File: workspace/go/tests/unit/user_service_test.go
/**
 * User Service Unit Tests
 */

package services

import (
	"errors"
	"testing"

	"github.com/nexusforge/api/internal/models"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/mock"
)

// Mock Repository
type MockUserRepository struct {
	mock.Mock
}

func (m *MockUserRepository) Create(user *models.User) error {
	args := m.Called(user)
	return args.Error(0)
}

func (m *MockUserRepository) FindByID(id uint) (*models.User, error) {
	args := m.Called(id)
	if args.Get(0) == nil {
		return nil, args.Error(1)
	}
	return args.Get(0).(*models.User), args.Error(1)
}

func (m *MockUserRepository) FindByEmail(email string) (*models.User, error) {
	args := m.Called(email)
	if args.Get(0) == nil {
		return nil, args.Error(1)
	}
	return args.Get(0).(*models.User), args.Error(1)
}

func (m *MockUserRepository) FindByUsername(username string) (*models.User, error) {
	args := m.Called(username)
	if args.Get(0) == nil {
		return nil, args.Error(1)
	}
	return args.Get(0).(*models.User), args.Error(1)
}

func (m *MockUserRepository) List(page, limit int) ([]*models.User, int64, error) {
	args := m.Called(page, limit)
	return args.Get(0).([]*models.User), args.Get(1).(int64), args.Error(2)
}

func (m *MockUserRepository) Update(user *models.User) error {
	args := m.Called(user)
	return args.Error(0)
}

func (m *MockUserRepository) Delete(id uint) error {
	args := m.Called(id)
	return args.Error(0)
}

func (m *MockUserRepository) UpdateLastLogin(id uint) error {
	args := m.Called(id)
	return args.Error(0)
}

// Mock Cache Manager
type MockCacheManager struct {
	mock.Mock
}

func (m *MockCacheManager) Get(key string, dest interface{}) error {
	args := m.Called(key, dest)
	return args.Error(0)
}

func (m *MockCacheManager) Set(key string, value interface{}, ttl any) error {
	args := m.Called(key, value, ttl)
	return args.Error(0)
}

func (m *MockCacheManager) Delete(key string) error {
	args := m.Called(key)
	return args.Error(0)
}

func (m *MockCacheManager) Exists(key string) (bool, error) {
	args := m.Called(key)
	return args.Bool(0), args.Error(1)
}

func (m *MockCacheManager) DeletePattern(pattern string) error {
	args := m.Called(pattern)
	return args.Error(0)
}

// Mock Logger
type MockLogger struct{}

func (m *MockLogger) Debug(msg string, fields ...interface{}) {}
func (m *MockLogger) Info(msg string, fields ...interface{})  {}
func (m *MockLogger) Warn(msg string, fields ...interface{})  {}
func (m *MockLogger) Error(msg string, fields ...interface{}) {}
func (m *MockLogger) Fatal(msg string, fields ...interface{}) {}

func TestCreateUser_Success(t *testing.T) {
	mockRepo := new(MockUserRepository)
	mockCache := new(MockCacheManager)
	mockLog := &MockLogger{}

	service := NewUserService(mockRepo, mockCache, mockLog)

	req := &models.CreateUserRequest{
		Email:    "test@example.com",
		Username: "testuser",
		Password: "password123",
	}

	mockRepo.On("FindByEmail", req.Email).Return(nil, nil)
	mockRepo.On("FindByUsername", req.Username).Return(nil, nil)
	mockRepo.On("Create", mock.AnythingOfType("*models.User")).Return(nil)

	user, err := service.CreateUser(req)

	assert.NoError(t, err)
	assert.NotNil(t, user)
	assert.Equal(t, req.Email, user.Email)
	assert.Equal(t, req.Username, user.Username)
	mockRepo.AssertExpectations(t)
}

func TestCreateUser_EmailExists(t *testing.T) {
	mockRepo := new(MockUserRepository)
	mockCache := new(MockCacheManager)
	mockLog := &MockLogger{}

	service := NewUserService(mockRepo, mockCache, mockLog)

	req := &models.CreateUserRequest{
		Email:    "existing@example.com",
		Username: "testuser",
		Password: "password123",
	}

	existingUser := &models.User{ID: 1, Email: req.Email}
	mockRepo.On("FindByEmail", req.Email).Return(existingUser, nil)

	user, err := service.CreateUser(req)

	assert.Error(t, err)
	assert.Nil(t, user)
	assert.Equal(t, "email already registered", err.Error())
	mockRepo.AssertExpectations(t)
}

func TestGetUserByID_FromCache(t *testing.T) {
	mockRepo := new(MockUserRepository)
	mockCache := new(MockCacheManager)
	mockLog := &MockLogger{}

	service := NewUserService(mockRepo, mockCache, mockLog)

	userID := uint(1)
	mockCache.On("Get", "user:1", mock.AnythingOfType("*models.UserResponse")).Return(nil)

	user, err := service.GetUserByID(userID)

	assert.NoError(t, err)
	assert.NotNil(t, user)
	mockCache.AssertExpectations(t)
}

func TestDeleteUser_Success(t *testing.T) {
	mockRepo := new(MockUserRepository)
	mockCache := new(MockCacheManager)
	mockLog := &MockLogger{}

	service := NewUserService(mockRepo, mockCache, mockLog)

	userID := uint(1)
	mockRepo.On("Delete", userID).Return(nil)
	mockCache.On("Delete", "user:1").Return(nil)

	err := service.DeleteUser(userID)

	assert.NoError(t, err)
	mockRepo.AssertExpectations(t)
	mockCache.AssertExpectations(t)
}

File: workspace/nodejs/tests/unit/user.service.test.ts
/**
 * User Service Unit Tests
 */

import { UserService } from '@/services/user.service';
import { prisma } from '@/utils/database';
import { cacheManager } from '@/utils/cache';
import { ConflictError, NotFoundError } from '@/utils/errors';

// Mock dependencies
jest.mock('@/utils/database');
jest.mock('@/utils/cache');

describe('UserService', () => {
  let userService: UserService;

  beforeEach(() => {
    userService = new UserService();
    jest.clearAllMocks();
  });

  describe('createUser', () => {
    it('should create a new user successfully', async () => {
      const userData = {
        email: 'test@example.com',
        username: 'testuser',
        password: 'SecurePass123!',
      };

      const mockUser = {
        id: 1,
        email: userData.email,
        username: userData.username,
        hashedPassword: 'hashed_password',
        isActive: true,
        isSuperuser: false,
        isEmailVerified: false,
        createdAt: new Date(),
        updatedAt: new Date(),
        lastLogin: null,
      };

      (prisma.user.findFirst as jest.Mock).mockResolvedValue(null);
      (prisma.user.create as jest.Mock).mockResolvedValue(mockUser);

      const result = await userService.createUser(userData);

      expect(result).toMatchObject({
        id: 1,
        email: userData.email,
        username: userData.username,
      });
      expect(result).not.toHaveProperty('hashedPassword');
    });

    it('should throw ConflictError if email already exists', async () => {
      const userData = {
        email: 'existing@example.com',
        username: 'testuser',
        password: 'SecurePass123!',
      };

      (prisma.user.findFirst as jest.Mock).mockResolvedValue({ id: 1 });

      await expect(userService.createUser(userData)).rejects.toThrow(ConflictError);
    });
  });

  describe('getUserById', () => {
    it('should return user from cache if available', async () => {
      const mockUser = {
        id: 1,
        email: 'test@example.com',
        username: 'testuser',
        isActive: true,
        isSuperuser: false,
        isEmailVerified: false,
        createdAt: new Date(),
        updatedAt: new Date(),
        lastLogin: null,
      };

      (cacheManager.get as jest.Mock).mockResolvedValue(mockUser);

      const result = await userService.getUserById(1);

      expect(result).toEqual(mockUser);
      expect(prisma.user.findUnique).not.toHaveBeenCalled();
    });

    it('should fetch user from database and cache it', async () => {
      const mockUser = {
        id: 1,
        email: 'test@example.com',
        username: 'testuser',
        hashedPassword: 'hashed_password',
        isActive: true,
        isSuperuser: false,
        isEmailVerified: false,
        createdAt: new Date(),
        updatedAt: new Date(),
        lastLogin: null,
      };

      (cacheManager.get as jest.Mock).mockResolvedValue(null);
      (prisma.user.findUnique as jest.Mock).mockResolvedValue(mockUser);
      (cacheManager.set as jest.Mock).mockResolvedValue(true);

      const result = await userService.getUserById(1);

      expect(result).toMatchObject({
        id: 1,
        email: 'test@example.com',
        username: 'testuser',
      });
      expect(result).not.toHaveProperty('hashedPassword');
      expect(cacheManager.set).toHaveBeenCalled();
    });

    it('should throw NotFoundError if user does not exist', async () => {
      (cacheManager.get as jest.Mock).mockResolvedValue(null);
      (prisma.user.findUnique as jest.Mock).mockResolvedValue(null);

      await expect(userService.getUserById(999)).rejects.toThrow(NotFoundError);
    });
  });

  describe('updateUser', () => {
    it('should update user and invalidate cache', async () => {
      const userId = 1;
      const updateData = {
        username: 'newusername',
      };

      const mockUser = {
        id: userId,
        email: 'test@example.com',
        username: updateData.username,
        hashedPassword: 'hashed_password',
        isActive: true,
        isSuperuser: false,
        isEmailVerified: false,
        createdAt: new Date(),
        updatedAt: new Date(),
        lastLogin: null,
      };

      (prisma.user.findFirst as jest.Mock).mockResolvedValue(null);
      (prisma.user.update as jest.Mock).mockResolvedValue(mockUser);
      (cacheManager.delete as jest.Mock).mockResolvedValue(true);

      const result = await userService.updateUser(userId, updateData);

      expect(result.username).toBe(updateData.username);
      expect(cacheManager.delete).toHaveBeenCalledWith(`user:${userId}`);
    });
  });

  describe('deleteUser', () => {
    it('should soft delete user and invalidate cache', async () => {
      const userId = 1;

      const mockUser = {
        id: userId,
        email: 'test@example.com',
        username: 'testuser',
        hashedPassword: 'hashed_password',
        isActive: false,
        isSuperuser: false,
        isEmailVerified: false,
        createdAt: new Date(),
        updatedAt: new Date(),
        lastLogin: null,
      };

      (prisma.user.update as jest.Mock).mockResolvedValue(mockUser);
      (cacheManager.delete as jest.Mock).mockResolvedValue(true);

      await userService.deleteUser(userId);

      expect(prisma.user.update).toHaveBeenCalledWith({
        where: { id: userId },
        data: { isActive: false },
      });
      expect(cacheManager.delete).toHaveBeenCalledWith(`user:${userId}`);
    });
  });
});
